### PR TITLE
OpenStack: Remove FIPs of LBs created by cloud-provider

### PR DESCRIPTION
**Pull request by dulek**
To implement `LoadBalancer` `Services` OpenStack Cloud Provider creates an
LB in Octavia for each of them. Those LBs have floating IPs associated
to the LB VIPs. As this is handled by the Cloud Provider itself,
Octavia's cascade delete will not remove those FIPs and we need to take
care of that ourselves.

This commit adds deleting of the FIPs associated to the LBs on
destroying cluster. In order to do that correctly it was required to
split deleting routers into two separate functions and make sure that
FIP detach and actual removal of the routers is happening after all the
delete functions finished (meaning that LBs are gone too). This is
because if FIP detach happens before we lose information about which FIP
was attached to which LB effectively preventing us from handling
deletion of them.

Only the function detaching the subnets from the routers is running as
part of main deletion step now.

### CODE REVIEW

File: pkg/destroy/openstack/openstack.go [hunk 1]
```diff
@@ -449,31 +457,108 @@ func updateFips(allFIPs []floatingips.FloatingIP, opts *clientconfig.ClientOpts,
...
 	}
 
-	allRouters, err := routers.ExtractRouters(allPages)
+	numberToClear := len(allRouters)
+	numberCleaned := 0
+	for _, router := range allRouters {
+		_, err = removeRouterInterfaces(conn, filter, router, logger)
+		if err != nil {
+			continue
+		}
```
Comment (pierreprinetti) [inline pkg/destroy/openstack/openstack.go:539]: Do you want to `logger.Debug` the error instead of silently ignoring?

Comment (dulek) [reply pkg/destroy/openstack/openstack.go:539]: It was like this before [1], but yeah.

[1] https://github.com/openshift/installer/pull/5050/files/d87b0c6138b17303dd4e3e195cf15d7e735e8629#diff-e480f6c0393b6788fe8352f4f8bdc63f021f9e84de632a21607aab8b414c6e3fL512-L515

File: pkg/destroy/openstack/openstack.go [hunk 2]
```diff
@@ -125,6 +125,14 @@ func (o *ClusterUninstaller) Run() error {
 		o.Logger.Debugf("goroutine %v complete", res)
 	}
 
+	// we want to remove routers as the last thing as it requires detaching the
+	// FIPs and that will cause it impossible to track which FIPs are tied to
+	// LBs being deleted.
+	err = deleteRouterRunner(opts, o.Filter, o.Logger)
```
Comment (mandre) [inline pkg/destroy/openstack/openstack.go:131]: Is there a risk the router is going to block the completion of any other resource deletion? If so you'll never reach this code. I think not but just want to make sure...

Comment (mandre) [reply pkg/destroy/openstack/openstack.go:131]: Hmm, after looking at CI this is probably what's happening:
```
 INFO[2021-07-02T11:49:09Z] Deprovisioning cluster ...
{"component":"entrypoint","file":"prow/entrypoint/run.go:165","func":"k8s.io/test-infra/prow/entrypoint.Options.ExecuteProcess","level":"error","msg":"Process did not finish before 45m0s timeout","severity":"error","time":"2021-07-02T11:39:06Z"} 
```

Comment (mandre) [reply pkg/destroy/openstack/openstack.go:131]: If that's really what's happening you need to extract the deletion of the blocking resources to your second runner.

Comment (dulek) [reply pkg/destroy/openstack/openstack.go:131]: Yep, I'll need to analyze that. :/

Comment (dulek) [reply pkg/destroy/openstack/openstack.go:131]: Oh, seems like if it fails to remove an interface from the router it won't keep retrying so interface is left and blocks subnet removal. Easy fix but I need to check if it works.

File: pkg/destroy/openstack/openstack.go [hunk 3]
```diff
@@ -449,31 +457,109 @@ func updateFips(allFIPs []floatingips.FloatingIP, opts *clientconfig.ClientOpts,
...
+
+	allRouters, err := getRouters(opts, filter, logger)
 	if err != nil {
 		logger.Error(err)
 		return false, nil
 	}
 
-	allRouters, err := routers.ExtractRouters(allPages)
+	numberToClear := len(allRouters)
+	numberCleaned := 0
```
[the remark this answers is not in the document]
Comment (dulek) [reply pkg/destroy/openstack/openstack.go:534]: I wrote that initially and changed it. My brain is playing with me.

File: pkg/destroy/openstack/openstack.go [hunk 4]
```diff
@@ -449,31 +457,112 @@ func updateFips(allFIPs []floatingips.FloatingIP, opts *clientconfig.ClientOpts,
...
+	numberCleared := 0
+	for _, router := range allRouters {
+		removed, err := removeRouterInterfaces(conn, filter, router, logger)
+		if err != nil {
+			logger.Debug(err)
+			continue
+		}
+		if !removed {
+			continue
+		}
```
Comment (dulek) [inline pkg/destroy/openstack/openstack.go:543]: @mandre: This was the problem, function's result was ignored. This piece seems ugly, I'm happy to hear suggestions how to code it better.

**Comment by dulek** [comment]
Looks like it worked!

**Review by mandre — COMMENTED** [review]
/lgtm

**Comment by mandre** [comment]
Nicely done @dulek 
/approve

**Pull request merged**